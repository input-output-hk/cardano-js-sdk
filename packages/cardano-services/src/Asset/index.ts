export * from './AssetHttpService';
export * from './CardanoTokenRegistry';
export * from './StubTokenMetadataService';
export * from './DbSyncNftMetadataService';
export * from './DbSyncAssetProvider';
export * from './TypeOrmNftMetadataService';
export * from './types';
