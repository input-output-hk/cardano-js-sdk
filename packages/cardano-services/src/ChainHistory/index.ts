export * from './ChainHistoryHttpService.js';
export * from './DbSyncChainHistory/index.js';
