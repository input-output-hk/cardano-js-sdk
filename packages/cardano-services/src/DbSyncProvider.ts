import { HealthCheckResponse, Provider } from '@cardano-sdk/core';
import { Pool } from 'pg';

const HEALTH_CHECK_QUERY = 'SELECT 1';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyArgs = any[];

export const DbSyncProvider = <
  T extends (abstract new (...args: AnyArgs) => {}) | (new (...args: AnyArgs) => {}) = { new (): {} }
>(
  BaseClass?: T
) => {
  abstract class Mixin extends (BaseClass || Object) implements Provider {
    public db: Pool;

    constructor(...args: AnyArgs) {
      const [db, ...baseArgs] = [...args];

      super(...baseArgs);

      this.db = db;
    }

    public async healthCheck(): Promise<HealthCheckResponse> {
      const result = await this.db.query(HEALTH_CHECK_QUERY);
      return { ok: !!result.rowCount };
    }
  }

  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
    ? A
    : T extends new (...baseArgs: infer A) => {}
    ? A
    : never;
  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
    ? I
    : T extends new (...baseArgs: AnyArgs) => infer I
    ? I
    : never;
  type ReturnedType = BaseInstance & { db: Pool; healthCheck: () => Promise<HealthCheckResponse> };

  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
    ? new (db: Pool, ...args: BaseArgs) => ReturnedType
    : abstract new (db: Pool, ...args: BaseArgs) => ReturnedType) & {
    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
  };
};

export type DbSyncProvider<T> = T extends new (...args: AnyArgs) => infer O ? O : never;
