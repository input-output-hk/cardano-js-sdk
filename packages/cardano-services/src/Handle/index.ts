export * from './HandleHttpService';
export * from './TypeOrmHandleProvider';
