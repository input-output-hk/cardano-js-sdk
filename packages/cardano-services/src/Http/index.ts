export * from './HttpServer';
export * from './HttpService';
export * from './types';
export * from './util';
