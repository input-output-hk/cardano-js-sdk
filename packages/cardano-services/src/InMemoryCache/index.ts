export * from './InMemoryCache';
export * from './NoCache';
export * from './defaults';
