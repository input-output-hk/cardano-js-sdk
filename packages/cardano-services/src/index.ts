export * from './Asset';
export * from './Http';
export * from './InMemoryCache';
export * from './Program';
export * from './RunnableModule';
export * from './StakePool';
export * from './TxSubmit';
export * from './Utxo';
export * from './ChainHistory';
export * from './Rewards';
export * as errors from './errors';
export * as util from './util';
