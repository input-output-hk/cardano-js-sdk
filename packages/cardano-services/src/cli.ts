#!/usr/bin/env node

/* eslint-disable no-console */
/* eslint-disable complexity */
/* eslint-disable sonarjs/cognitive-complexity */
/* eslint-disable unicorn/no-nested-ternary */
import {
  ALLOWED_ORIGINS_DEFAULT,
  AvailableNetworks,
  BLOCKFROST_WORKER_API_URL_DEFAULT,
  BlockfrostWorkerArgs,
  BlockfrostWorkerOptionDescriptions,
  CACHE_TTL_DEFAULT,
  CREATE_SCHEMA_DEFAULT,
  DISABLE_DB_CACHE_DEFAULT,
  DISABLE_STAKE_POOL_METRIC_APY_DEFAULT,
  DROP_SCHEMA_DEFAULT,
  DRY_RUN_DEFAULT,
  HTTP_SERVER_API_URL_DEFAULT,
  PAGINATION_PAGE_SIZE_LIMIT_DEFAULT,
  PARALLEL_MODE_DEFAULT,
  PARALLEL_TXS_DEFAULT,
  POLLING_CYCLE_DEFAULT,
  Programs,
  ProviderServerArgs,
  ProviderServerOptionDescriptions,
  SCAN_INTERVAL_DEFAULT,
  ServiceNames,
  TX_WORKER_API_URL_DEFAULT,
  TxWorkerArgs,
  TxWorkerOptionDescriptions,
  USE_BLOCKFROST_DEFAULT,
  USE_QUEUE_DEFAULT,
  availableNetworks,
  connectionStringFromArgs,
  loadAndStartTxWorker,
  loadBlockfrostWorker,
  loadProviderServer,
  stringOptionToBoolean
} from './Program';
import { Command, Option } from 'commander';
import { DB_CACHE_TTL_DEFAULT } from './InMemoryCache';
import {
  DEFAULT_TOKEN_METADATA_CACHE_TTL,
  DEFAULT_TOKEN_METADATA_REQUEST_TIMEOUT,
  DEFAULT_TOKEN_METADATA_SERVER_URL
} from './Asset';
import { EPOCH_POLL_INTERVAL_DEFAULT } from './util';
import { HttpServer } from './Http';
import { URL } from 'url';
import { dbCacheValidator } from './util/validators';
import { withCommonOptions, withOgmiosOptions, withPostgresOptions, withRabbitMqOptions } from './Program/options/';
import fs from 'fs';
import onDeath from 'death';
import path from 'path';

const copiedPackageJsonPath = path.join(__dirname, 'original-package.json');
// Exists in dist/, doesn't exist when run with ts-node
const packageJsonPath = fs.existsSync(copiedPackageJsonPath)
  ? copiedPackageJsonPath
  : path.join(__dirname, '../package.json');
const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));

process.on('unhandledRejection', (reason) => {
  // To be handled by 'onDeath'
  throw reason;
});

console.log('Cardano Services CLI');

const program = new Command('cardano-services');

program.version(packageJson.version);

const runServer = async (message: string, loadServer: () => Promise<HttpServer>) => {
  try {
    console.log(message);
    const server = await loadServer();

    await server.initialize();
    await server.start();

    onDeath(async () => {
      await server.shutdown();
      process.exit(1);
    });
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
};

withCommonOptions(
  withOgmiosOptions(
    withPostgresOptions(
      withRabbitMqOptions(
        program
          .command('start-provider-server')
          .description('Start the Provider Server')
          .argument('[serviceNames...]', `List of services to attach: ${Object.values(ServiceNames).toString()}`)
      )
    )
  ),
  {
    apiUrl: HTTP_SERVER_API_URL_DEFAULT
  }
)
  .addOption(
    new Option(
      '--service-names <serviceNames>',
      `List of services to attach: ${Object.values(ServiceNames).toString()}`
    )
      .env('SERVICE_NAMES')
      .argParser((names) => names.split(',') as ServiceNames[])
  )
  .addOption(
    new Option('--allowed-origins <allowedOrigins>', ProviderServerOptionDescriptions.AllowedOrigins)
      .env('ALLOWED_ORIGINS')
      .default(ALLOWED_ORIGINS_DEFAULT)
      .argParser((originsList) => originsList.split(',') as string[])
  )
  .addOption(
    new Option(
      '--cardano-node-config-path <cardanoNodeConfigPath>',
      ProviderServerOptionDescriptions.CardanoNodeConfigPath
    ).env('CARDANO_NODE_CONFIG_PATH')
  )
  .addOption(
    new Option('--db-cache-ttl <dbCacheTtl>', ProviderServerOptionDescriptions.DbCacheTtl)
      .env('DB_CACHE_TTL')
      .default(DB_CACHE_TTL_DEFAULT)
      .argParser(dbCacheValidator)
  )
  .addOption(
    new Option('--disable-db-cache <true/false>', ProviderServerOptionDescriptions.DisableDbCache)
      .env('DISABLE_DB_CACHE')
      .default(DISABLE_DB_CACHE_DEFAULT)
      .argParser((disableDbCache) =>
        stringOptionToBoolean(disableDbCache, Programs.ProviderServer, ProviderServerOptionDescriptions.DisableDbCache)
      )
  )
  .addOption(
    new Option(
      '--disable-stake-pool-metric-apy <true/false>',
      ProviderServerOptionDescriptions.DisableStakePoolMetricApy
    )
      .env('DISABLE_STAKE_POOL_METRIC_APY')
      .default(DISABLE_STAKE_POOL_METRIC_APY_DEFAULT)
      .argParser((disableStakePoolMetricApy) =>
        stringOptionToBoolean(
          disableStakePoolMetricApy,
          Programs.ProviderServer,
          ProviderServerOptionDescriptions.DisableStakePoolMetricApy
        )
      )
  )
  .addOption(
    new Option('--epoch-poll-interval <epochPollInterval>', ProviderServerOptionDescriptions.EpochPollInterval)
      .env('EPOCH_POLL_INTERVAL')
      .default(EPOCH_POLL_INTERVAL_DEFAULT)
      .argParser((interval) => Number.parseInt(interval, 10))
  )
  .addOption(
    new Option(
      '--token-metadata-server-url <tokenMetadataServerUrl>',
      ProviderServerOptionDescriptions.TokenMetadataServerUrl
    )
      .env('TOKEN_METADATA_SERVER_URL')
      .default(DEFAULT_TOKEN_METADATA_SERVER_URL)
      .argParser((url) => new URL(url).toString())
  )
  .addOption(
    new Option(
      '--token-metadata-cache-ttl <tokenMetadataCacheTTL>',
      ProviderServerOptionDescriptions.TokenMetadataCacheTtl
    )
      .env('TOKEN_METADATA_CACHE_TTL')
      .default(DEFAULT_TOKEN_METADATA_CACHE_TTL)
      .argParser(dbCacheValidator)
  )
  .addOption(
    new Option(
      '--token-metadata-request-timeout <tokenMetadataRequestTimeout>',
      ProviderServerOptionDescriptions.PaginationPageSizeLimit
    )
      .env('TOKEN_METADATA_REQUEST_TIMEOUT')
      .default(DEFAULT_TOKEN_METADATA_REQUEST_TIMEOUT)
      .argParser((interval) => Number.parseInt(interval, 10))
  )
  .addOption(
    new Option('--use-blockfrost <true/false>', ProviderServerOptionDescriptions.UseBlockfrost)
      .env('USE_BLOCKFROST')
      .default(USE_BLOCKFROST_DEFAULT)
      .argParser((useBlockfrost) =>
        stringOptionToBoolean(useBlockfrost, Programs.ProviderServer, ProviderServerOptionDescriptions.UseBlockfrost)
      )
  )
  .addOption(
    new Option('--use-queue <true/false>', ProviderServerOptionDescriptions.UseQueue)
      .env('USE_QUEUE')
      .default(USE_QUEUE_DEFAULT)
      .argParser((useQueue) =>
        stringOptionToBoolean(useQueue, Programs.ProviderServer, ProviderServerOptionDescriptions.UseQueue)
      )
  )
  .addOption(
    new Option(
      '--pagination-page-size-limit <paginationPageSizeLimit>',
      ProviderServerOptionDescriptions.PaginationPageSizeLimit
    )
      .env('PAGINATION_PAGE_SIZE_LIMIT')
      .default(PAGINATION_PAGE_SIZE_LIMIT_DEFAULT)
      .argParser((interval) => Number.parseInt(interval, 10))
  )
  .action(async (serviceNames: ServiceNames[], args: ProviderServerArgs) =>
    runServer('Provider server', () =>
      loadProviderServer({
        ...args,
        postgresConnectionString: connectionStringFromArgs(args),
        // Setting the service names via env variable takes preference over command line argument
        serviceNames: args.serviceNames ? args.serviceNames : serviceNames
      })
    )
  );

withCommonOptions(
  withOgmiosOptions(withRabbitMqOptions(program.command('start-worker').description('Start RabbitMQ worker'))),
  {
    apiUrl: TX_WORKER_API_URL_DEFAULT
  }
)
  .addOption(
    new Option('--parallel <true/false>', TxWorkerOptionDescriptions.Parallel)
      .env('PARALLEL')
      .default(PARALLEL_MODE_DEFAULT)
      .argParser((parallel) =>
        stringOptionToBoolean(parallel, Programs.RabbitmqWorker, TxWorkerOptionDescriptions.Parallel)
      )
  )
  .addOption(
    new Option('--parallel-txs <parallelTxs>', TxWorkerOptionDescriptions.ParallelTxs)
      .env('PARALLEL_TXS')
      .default(PARALLEL_TXS_DEFAULT)
      .argParser((parallelTxs) => Number.parseInt(parallelTxs, 10))
  )
  .addOption(
    new Option('--polling-cycle <pollingCycle>', TxWorkerOptionDescriptions.PollingCycle)
      .env('POLLING_CYCLE')
      .default(POLLING_CYCLE_DEFAULT)
      .argParser((pollingCycle) => Number.parseInt(pollingCycle, 10))
  )
  .action(async (args: TxWorkerArgs) => {
    // eslint-disable-next-line no-console
    console.log(`RabbitMQ transactions worker: ${args.parallel ? 'parallel' : 'serial'} mode`);
    const txWorker = await loadAndStartTxWorker(args);
    onDeath(async () => {
      await txWorker.shutdown();
      process.exit(1);
    });
  });

withCommonOptions(
  withPostgresOptions(program.command('start-blockfrost-worker').description('Start the Blockfrost worker')),
  {
    apiUrl: BLOCKFROST_WORKER_API_URL_DEFAULT
  }
)
  .addOption(
    new Option('--blockfrost-api-file <blockfrostApiFile>', BlockfrostWorkerOptionDescriptions.BlockfrostApiFile)
      .env('BLOCKFROST_API_FILE')
      .conflicts('blockfrostApiKey')
  )
  .addOption(
    new Option('--blockfrost-api-key <blockfrostApiKey>', BlockfrostWorkerOptionDescriptions.BlockfrostApiKey).env(
      'BLOCKFROST_API_KEY'
    )
  )
  .addOption(
    new Option('--cache-ttl <cacheTtl>', BlockfrostWorkerOptionDescriptions.CacheTTL)
      .env('CACHE_TTL')
      .default(CACHE_TTL_DEFAULT)
      .argParser((interval) => Number.parseInt(interval, 10))
  )
  .addOption(
    new Option('--create-schema <true/false>', BlockfrostWorkerOptionDescriptions.CreateSchema)
      .env('CREATE_SCHEMA')
      .default(CREATE_SCHEMA_DEFAULT)
      .argParser((createSchema) =>
        stringOptionToBoolean(createSchema, Programs.BlockfrostWorker, BlockfrostWorkerOptionDescriptions.CreateSchema)
      )
  )
  .addOption(
    new Option('--drop-schema <true/false>', BlockfrostWorkerOptionDescriptions.DropSchema)
      .env('DROP_SCHEMA')
      .default(DROP_SCHEMA_DEFAULT)
      .argParser((dropSchema) =>
        stringOptionToBoolean(dropSchema, Programs.BlockfrostWorker, BlockfrostWorkerOptionDescriptions.DropSchema)
      )
  )
  .addOption(
    new Option('--dry-run <true/false>', BlockfrostWorkerOptionDescriptions.DryRun)
      .env('DRY_RUN')
      .default(DRY_RUN_DEFAULT)
      .argParser((dryRun) =>
        stringOptionToBoolean(dryRun, Programs.BlockfrostWorker, BlockfrostWorkerOptionDescriptions.DryRun)
      )
  )
  .addOption(
    new Option('--network <network>', BlockfrostWorkerOptionDescriptions.Network)
      .env('NETWORK')
      .argParser((network) => {
        if (availableNetworks.includes(network as AvailableNetworks)) return network;

        throw new Error(`Unknown network: ${network}`);
      })
  )
  .addOption(
    new Option('--scan-interval <scanInterval>', BlockfrostWorkerOptionDescriptions.ScanInterval)
      .env('SCAN_INTERVAL')
      .default(SCAN_INTERVAL_DEFAULT)
      .argParser((interval) => Number.parseInt(interval, 10))
  )
  .action(async (args: BlockfrostWorkerArgs) =>
    runServer('Blockfrost worker', () =>
      loadBlockfrostWorker({ ...args, postgresConnectionString: connectionStringFromArgs(args) })
    )
  );

if (process.argv.slice(2).length === 0) {
  program.outputHelp();
  process.exit(1);
} else {
  program.parseAsync(process.argv).catch((error) => {
    // eslint-disable-next-line no-console
    console.error(error);
    process.exit(1);
  });
}
