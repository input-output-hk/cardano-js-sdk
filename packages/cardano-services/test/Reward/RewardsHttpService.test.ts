/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable max-len */
import { Cardano, ProviderError, ProviderFailure, RewardsProvider } from '@cardano-sdk/core';
import { CreateHttpProviderConfig, rewardsHttpProvider } from '@cardano-sdk/cardano-services-client';
import { DbSyncRewardsProvider, HttpServer, HttpServerConfig, RewardsHttpService } from '../../src';
import { INFO, createLogger } from 'bunyan';
import { Pool } from 'pg';
import { getPort } from 'get-port-please';
import { dummyLogger as logger } from 'ts-log';
import axios from 'axios';

const APPLICATION_JSON = 'application/json';
const UNSUPPORTED_MEDIA_STRING = 'Request failed with status code 415';
const APPLICATION_CBOR = 'application/cbor';
const BAD_REQUEST_STRING = 'Request failed with status code 400';

describe('RewardsHttpService', () => {
  let dbConnection: Pool;
  let httpServer: HttpServer;
  let rewardsProvider: DbSyncRewardsProvider;
  let service: RewardsHttpService;
  let port: number;
  let baseUrl: string;
  let clientConfig: CreateHttpProviderConfig<RewardsProvider>;
  let config: HttpServerConfig;

  beforeAll(async () => {
    port = await getPort();
    baseUrl = `http://localhost:${port}/rewards`;
    clientConfig = { baseUrl, logger: createLogger({ level: INFO, name: 'unit tests' }) };
    config = { listen: { port } };
    dbConnection = new Pool({ connectionString: process.env.POSTGRES_CONNECTION_STRING });
  });

  afterEach(async () => {
    jest.resetAllMocks();
  });

  describe('unhealthy RewardsProvider', () => {
    beforeEach(async () => {
      rewardsProvider = {
        healthCheck: jest.fn(() => Promise.resolve({ ok: false })),
        rewardAccountBalance: jest.fn(),
        rewardsHistory: jest.fn()
      } as unknown as DbSyncRewardsProvider;
    });

    it('should not throw during service create if the RewardsProvider is unhealthy', () => {
      expect(() => new RewardsHttpService({ logger, rewardsProvider })).not.toThrow(
        new ProviderError(ProviderFailure.Unhealthy)
      );
    });

    it('throws during service initialization if the RewardsProvider is unhealthy', async () => {
      service = new RewardsHttpService({ logger, rewardsProvider });
      httpServer = new HttpServer(config, { logger, services: [service] });
      await expect(httpServer.initialize()).rejects.toThrow(new ProviderError(ProviderFailure.Unhealthy));
    });
  });

  describe('healthy state', () => {
    beforeAll(async () => {
      rewardsProvider = new DbSyncRewardsProvider(dbConnection, logger);
      service = new RewardsHttpService({ logger, rewardsProvider });
      httpServer = new HttpServer(config, { logger, services: [service] });
      await httpServer.initialize();
      await httpServer.start();
    });

    afterAll(async () => {
      await dbConnection.end();
      await httpServer.shutdown();
    });

    describe('/health', () => {
      it('forwards the stakePoolSearchProvider health response', async () => {
        const res = await axios.post(`${baseUrl}/health`, {}, { headers: { 'Content-Type': APPLICATION_JSON } });
        expect(res.status).toBe(200);
        expect(res.data).toEqual({ ok: true });
      });
    });

    describe('/history', () => {
      const historyUrl = '/history';
      const rewardAddress = 'stake_test1uqfu74w3wh4gfzu8m6e7j987h4lq9r3t7ef5gaw497uu85qsqfy27';
      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect(
          (
            await axios.post(`${baseUrl}${historyUrl}`, {
              args: [
                {
                  epochs: {
                    lowerBound: 1,
                    upperBound: 14
                  },
                  rewardAccounts: [rewardAddress]
                }
              ]
            })
          ).status
        ).toEqual(200);
      });
      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(
            `${baseUrl}${historyUrl}`,
            {
              args: [
                {
                  rewardAccounts: [rewardAddress]
                }
              ]
            },
            { headers: { 'Content-Type': APPLICATION_CBOR } }
          );
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });

      it('returns 400 coded respons if the request is bad formed', async () => {
        try {
          await axios.post(`${baseUrl}${historyUrl}`, { args: [{ field: 'value' }] });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(400);
          expect(error.message).toBe(BAD_REQUEST_STRING);
        }
      });
    });

    describe('/account-balance', () => {
      const accountBalanceUrl = '/account-balance';
      const rewardAddress = 'stake_test1uqfu74w3wh4gfzu8m6e7j987h4lq9r3t7ef5gaw497uu85qsqfy27';
      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect(
          (
            await axios.post(`${baseUrl}${accountBalanceUrl}`, {
              args: [rewardAddress]
            })
          ).status
        ).toEqual(200);
      });
      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(
            `${baseUrl}${accountBalanceUrl}`,
            { args: [rewardAddress] },
            { headers: { 'Content-Type': APPLICATION_CBOR } }
          );
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });
      it('returns 400 coded respons if the request is bad formed', async () => {
        try {
          await axios.post(`${baseUrl}${accountBalanceUrl}`, { args: [{ address: 'asd' }] });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(400);
          expect(error.message).toBe(BAD_REQUEST_STRING);
        }
      });
    });

    describe('with rewardsHttpProvider', () => {
      let provider: RewardsProvider;
      beforeEach(() => {
        provider = rewardsHttpProvider(clientConfig);
      });
      const rewardAcc = Cardano.RewardAccount('stake_test1upd9j9rwxeu44xfxnrl6sqsswf9k60gcdjuy2gz6zyu2jmqyvn80c');
      describe('rewardAccountBalance', () => {
        it('returns address balance', async () => {
          const response = await provider.rewardAccountBalance(rewardAcc);
          expect(response).toMatchSnapshot();
        });
        it('returns address balance 0 when it has no rewards', async () => {
          const response = await provider.rewardAccountBalance(
            Cardano.RewardAccount('stake_test1uzxvhl83q8ujv2yvpy6n2krvpdlqqx28h7e9vsk6re43h3c3kufy6')
          );
          expect(response).toMatchSnapshot();
        });
      });
      describe('rewardsHistory', () => {
        it('returns rewards address history', async () => {
          const response = await provider.rewardsHistory({
            rewardAccounts: [rewardAcc]
          });
          expect(response).toMatchSnapshot();
        });
        it('returns no rewards address history for empty reward accounts', async () => {
          const response = await provider.rewardsHistory({
            rewardAccounts: []
          });
          expect(response).toMatchSnapshot();
        });
        it('returns address rewards history with epochs ', async () => {
          const accountWithRewardsAtEpoch76 = Cardano.RewardAccount(
            'stake_test1up32f2hrv5ytqk8ad6e4apss5zrrjjlrkjhrksypn5g08fqrqf9gr'
          );
          const response = await provider.rewardsHistory({
            epochs: {
              lowerBound: 75,
              upperBound: 76
            },
            rewardAccounts: [accountWithRewardsAtEpoch76]
          });
          expect(response).toMatchSnapshot();
        });
        it('returns rewards address history of the epochs filtered', async () => {
          const response = await provider.rewardsHistory({
            epochs: {
              lowerBound: 10
            },
            rewardAccounts: [rewardAcc]
          });
          expect(response).toMatchSnapshot();
        });
        it('returns rewards address history some of the epochs filter', async () => {
          const response = await provider.rewardsHistory({
            epochs: {
              upperBound: 10
            },
            rewardAccounts: [rewardAcc]
          });
          expect(response).toMatchSnapshot();
        });
      });
    });
  });
});
