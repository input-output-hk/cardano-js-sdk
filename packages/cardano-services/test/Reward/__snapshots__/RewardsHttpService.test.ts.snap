// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardAccountBalance returns address balance 0 when it has no rewards 1`] = `0n`;

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardAccountBalance returns address balance 1`] = `5000000000n`;

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardsHistory returns address rewards history with epochs  1`] = `
Map {
  "stake_test1up32f2hrv5ytqk8ad6e4apss5zrrjjlrkjhrksypn5g08fqrqf9gr" => Array [
    Object {
      "epoch": "76",
      "rewards": 500000000n,
    },
  ],
}
`;

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardsHistory returns no rewards address history for empty reward accounts 1`] = `Map {}`;

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardsHistory returns rewards address history 1`] = `
Map {
  "stake_test1upd9j9rwxeu44xfxnrl6sqsswf9k60gcdjuy2gz6zyu2jmqyvn80c" => Array [
    Object {
      "epoch": "161",
      "rewards": 5000000000n,
    },
  ],
}
`;

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardsHistory returns rewards address history of the epochs filtered 1`] = `
Map {
  "stake_test1upd9j9rwxeu44xfxnrl6sqsswf9k60gcdjuy2gz6zyu2jmqyvn80c" => Array [
    Object {
      "epoch": "161",
      "rewards": 5000000000n,
    },
  ],
}
`;

exports[`RewardsHttpService healthy state with rewardsHttpProvider rewardsHistory returns rewards address history some of the epochs filter 1`] = `Map {}`;
